### DISCUSSION TITLE
MySQL connection from Docker action?

### DISCUSSION THREAD

**Discussion by Guite** [community/community#26224] [Actions]
<p>Hello,</p>
<p>can I somehow pass a service (specifically a database connection) to a Docker action?</p>
<p>When I try to pass options like this:</p>
<pre><code>with:
    database_user: myUser
    ...
    database_port: "${{ job.services.mysql.ports['3306'] }}"
</code></pre>
<p>and use this information inside my entrypoint.sh file</p>
<p>I get the following error:</p>
<pre><code>mysql -h 127.0.0.1 --port 32768 -u myUser -pxy -e "CREATE DATABASE IF NOT EXISTS myDB;"

ERROR 2002 (HY000): Can't connect to MySQL server on '127.0.0.1' (115)
</code></pre>
<p>Is there any possibility to map the connection? Or do I need to setup MySQL inside my Docker container?</p>
<p>Thanks for your advice,</p>
<p>Axel</p>

**Comment by BrightRan** [comment]
<p>There are some of the possible causes of this error:</p>
<ol>
<li>
<p>Network failure especially if mysql database server is running on remote host.</p>
</li>
<li>
<p>No mysql server is running on the mentioned host.</p>
</li>
<li>
<p>Firewall blocking TCP-IP connection or other related reasons.</p>
</li>
</ol>
<p>You can try using ping command to test the client-server connectivity, such as “ <strong>ping 127.0.0.1</strong> ”.</p>
<p>The following are two Docker container actions for MySQL and docs for MySQL image, maybe you can reference them:</p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmarketplace%2Factions%2Fsetup-mysql&amp;data=02%7C01%7Cv-brira%40microsoft.com%7C3b7a0c17b4ef40ad111708d777c970c8%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637109579369391659&amp;sdata=7V8MHQ9VvxBymknidItGSrcU8aI7gicp%2BiKpiRBcWAU%3D&amp;reserved=0" rel="nofollow noopener">Setup MySQL</a></p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmarketplace%2Factions%2Fstart-mysql&amp;data=02%7C01%7Cv-brira%40microsoft.com%7C3b7a0c17b4ef40ad111708d777c970c8%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637109579369391659&amp;sdata=5RiquTopNrOfH161aHkaA5t6fX4Uiox49oRjRbelcBc%3D&amp;reserved=0" rel="nofollow noopener">Start MySQL</a></p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhub.docker.com%2F_%2Fmysql&amp;data=02%7C01%7Cv-brira%40microsoft.com%7C3b7a0c17b4ef40ad111708d777c970c8%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637109579369401651&amp;sdata=2Hr7fcXxOxth8Fx144k%2BKnyGkRHaqYkmYg8jG3XSsVA%3D&amp;reserved=0" rel="nofollow noopener">https://hub.docker.com/_/mysql</a></p>

**Comment by Guite** [comment]
<p>I already have a MySQL server setup like this:</p>
<pre><code>services:
      mysql:
        image: mysql:5.7
        env:
          MYSQL_USER: myUser
          MYSQL_PASSWORD: myPass
          MYSQL_DATABASE: myDB
          MYSQL_ROOT_PASSWORD: myPass
        ports:
          - 3306
        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
</code></pre>
<p>But my special use case is that I do not want to use it directly in the workflow, but within a docker action used by the workflow. I will try to use the default mysql available in virtual environments instead.</p>

**Comment by Guite** [comment]
<p>This thread at stack overflow contains several possible solutions:</p>
<p><a href="https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach/52858101#52858101" rel="nofollow noopener">https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach/52858101#52858101</a></p>
<p>Is it possible with GitHub actions to call a docker image in host mode by setting/appending</p>
<pre><code>--network=host or --net=host
</code></pre>
<p>?</p>

**Comment by dakale** [comment, marked as answer]
<p>All containers (job, service, actions) get attached to the same user defined bridge network on the host, meaning all the containers can reach each other over that network, not via the host’s localhost networking. From inside your action, you should be able to connect to mysql with hostname mysql (comes from the name of your service) and the default port </p>

**Comment by Guite** [comment]
<p>Thank you very much! I’d never found this by myself.</p>

**Comment by Nosferican** [comment]
<p>If I wanted to connect to a remote webdriver from inside a GitHub Action,</p>
<p>after running</p>
<pre><code>docker run -d -p 4444:4444 --name selenium selenium/standalone-chrome
</code></pre>
<p>what address should I use rather than </p>
<pre><code>localhost:4444
</code></pre>

**Comment by dakale** [comment]
<p>You could run the webdriver as a service container, in which case youd do something like:</p>
<pre><code>services:
  selenium:
    image: selenium/standalone-chrome
</code></pre>
<p> (see <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idservices" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idservices</a>)</p>
<p>That way a docker network will be created, and from inside a container action, you can connect to it via the hostname “selenium” (comes from the service name). So youd just use hostname=selenium, port=4444</p>
<p>Note that in the above example, no port mapping were specified, thats because for user defined bridge networks in docker, all ports are accesible by default, so the application inside the container only needs to actually be listening on 4444</p>
<p>…</p>
<p>We dont currently support adding docker-specific options, like “network”, when running container actions so unfortunately you wouldnt be able to create a network, and run your action attached to that network specifically (something like docker run --network …)</p>
<p>Of course if you are using a javascript action that would be running on the host, so you could connect via localhost (if you ran the selenium container with --network host)</p>
<p>Hope that helps</p>

**Comment by zachguo** [comment]
<p>What if we have a custom docker container and cannot run it as a service? How can we connect to it?</p>
<pre><code>docker run -d -p 8378:8378 customcontainer:$GITHUB_SHA
curl --retry 10 --retry-delay 5 --retry-connrefused -v http://localhost:8378
</code></pre>
